### PR TITLE
Fix WebGL Conv2d for NCHW format

### PR DESCRIPTION
**Pull request by Linchenn**
Based on the findings from https://github.com/tensorflow/tfjs/issues/6308#issuecomment-1107033833, there are [3 branches](https://github.com/tensorflow/tfjs/blob/fda7d7b1014ef89d61d6329e8048da0fec8ed03e/tfjs-backend-webgl/src/kernels/Conv2D.ts#L39-L49) to compute Conv2d on WebGL. The first branch and the second branch are problematic for `NCHW` format, so this PR does the following:
1. The first branch does not consider the `NCHW` format, so this PR adds a condition for it to exclude computation for `NCHW` format. Whether this branch is supposed to support `NCHW` format deserves some further research.
2. The second branch is supposed to transpose the result of matrix multiplication to be `NCHW` format, instead of simply reshaping it. This PR fixes it.

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6340)

### CODE REVIEW

File: tfjs-backend-webgl/src/kernels/Conv2D.ts [hunk 1]
```diff
@@ -39,7 +39,8 @@ export function conv2d(
   if (convInfo.filterHeight === 1 && convInfo.filterWidth === 1 &&
       convInfo.dilationHeight === 1 && convInfo.dilationWidth === 1 &&
       convInfo.strideHeight === 1 && convInfo.strideWidth === 1 &&
-      (convInfo.padInfo.type === 'SAME' || convInfo.padInfo.type === 'VALID')) {
+      (convInfo.padInfo.type === 'SAME' || convInfo.padInfo.type === 'VALID') &&
+      $dataFormat === 'channelsLast') {
```
Comment (qjia7) [inline tfjs-backend-webgl/src/kernels/Conv2D.ts:43]: For 1x1 conv2d branch,
if the data format is NHWC, it becomes `xReshaped` * `filterReshaped` with `transposeA` = `false` and `transposeB` = `false`.
If the data format is NCHW, it becomes
`filterReshaped` * `xReshaped` with `transposeA` = `true` and `transposeB` = `false`.
So you still can go to the `conv2dByMatMul` path with some changes.

**Review by qjia7 — COMMENTED** [review]
@Linchenn Can you add a test to verify the [third path ](https://github.com/tensorflow/tfjs/blob/master/tfjs-backend-webgl/src/kernels/Conv2D.ts#L47) to see if it works correctly?

File: tfjs-backend-webgl/src/kernels/Conv2D.ts (hunk 1 above)
Comment (Linchenn) [reply tfjs-backend-webgl/src/kernels/Conv2D.ts]: Thank you for your idea about the 1x1 conv2d branch! Just added some reshaping and transposing logics to support the `NCHW` format. Also, added some basic tests for `NCHW` format and a test for `NCHW` format with batch=2.

We already have a [test](https://github.com/tensorflow/tfjs/blob/ff9910e898735e0a7ebe8f5e214bf3e872d4b7aa/tfjs-core/src/ops/conv2d_test.ts#L340-L359) of `NCHW` format for the third path.

**Review by Linchenn — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6340)* status: 0 of 1 approvals obtained (waiting on @pyu10055 and @qjia7)

___
*[`tfjs-backend-webgl/src/kernels/Conv2D_impl.ts` line 261 at r1](https://reviewable.io/reviews/tensorflow/tfjs/6340#-N0WKVMr9YlliXpR2iyy:-N0hXTIc8R7w73ENjOyg:b1mubt8) ([raw file](https://github.com/tensorflow/tfjs/blob/bbafbf612a8705cfe5286c6bd3798ea5c992d89a/tfjs-backend-webgl/src/kernels/Conv2D_impl.ts#L261)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

For NCHW conv2d, the bias is also in the NCHW format, it will require transpose if the matmul computation result is in NHWC format.
</blockquote></details>

Referring to [bias_add op in Python](https://www.tensorflow.org/api_docs/python/tf/nn/bias_add) involved in fusedConv2d, the bias is supposed to be a 1-D tensor aligned with the output channels.

I added asserts to check the shape of the bias to be a 1-D tensor or a scalar (the scalar could be broadcasted to be a 1-D tensor) and added tests to check conv2d computation with bias.

**Review by Linchenn — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6340)* status: 0 of 1 approvals obtained (waiting on @pyu10055 and @qjia7)

*[a discussion](https://reviewable.io/reviews/tensorflow/tfjs/6340#-N0m0Gsj8c_TjvMZXbT_:-N0m0Gsj8c_TjvMZXbTa:bmp857z) (no related file):*
In WebGL's [Conv2d_impl.ts](https://github.com/tensorflow/tfjs/blob/fda7d7b1014ef89d61d6329e8048da0fec8ed03e/tfjs-backend-webgl/src/kernels/Conv2D_impl.ts), both [conv2dByMatMul](https://github.com/tensorflow/tfjs/blob/fda7d7b1014ef89d61d6329e8048da0fec8ed03e/tfjs-backend-webgl/src/kernels/Conv2D_impl.ts#L44) and [conv2dWithIm2Row](https://github.com/tensorflow/tfjs/blob/fda7d7b1014ef89d61d6329e8048da0fec8ed03e/tfjs-backend-webgl/src/kernels/Conv2D_impl.ts#L184) take `bias` and `preluActivationWeights` as parameters. 

`bias` is supposed to be a 1-D tensor, so the dataFormat would not affect it. Added some tests in this PR.

`preluActivationWeights`, as defined in [fusedConv2d](https://github.com/tensorflow/tfjs/blob/fda7d7b1014ef89d61d6329e8048da0fec8ed03e/tfjs-core/src/ops/fused/conv2d.ts#L91), the shape could be similar with the input, which means it could have multiple dimensions and `dataFormat` may affect it, while the both functions do not consider `NCHW` format for it, which could be problematic:

>  @param preluActivationWeights Tensor of prelu weights to be applied as part of a `prelu` activation, typically the same shape as `x`.

In conclusion, `preluActivationWeights` deserves more research to check the correctness for `NCHW` format in WebGL, but this is unrelated to this PR, so I opened [a new ISSUE](https://github.com/tensorflow/tfjs/issues/6354) to track.

---

**Review by pyu10055 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6340)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @Linchenn and @qjia7)

*[a discussion](https://reviewable.io/reviews/tensorflow/tfjs/6340#-N0m0Gsj8c_TjvMZXbT_:-N0mGGrH71c8WuNSj8ic:bb8vsoh) (no related file):*
<details><summary><i>Previously, Linchenn wrote…</i></summary><blockquote>

In WebGL's [Conv2d_impl.ts](https://github.com/tensorflow/tfjs/blob/fda7d7b1014ef89d61d6329e8048da0fec8ed03e/tfjs-backend-webgl/src/kernels/Conv2D_impl.ts), both [conv2dByMatMul](https://github.com/tensorflow/tfjs/blob/fda7d7b1014ef89d61d6329e8048da0fec8ed03e/tfjs-backend-webgl/src/kernels/Conv2D_impl.ts#L44) and [conv2dWithIm2Row](https://github.com/tensorflow/tfjs/blob/fda7d7b1014ef89d61d6329e8048da0fec8ed03e/tfjs-backend-webgl/src/kernels/Conv2D_impl.ts#L184) take `bias` and `preluActivationWeights` as parameters. 

`bias` is supposed to be a 1-D tensor, so the dataFormat would not affect it. Added some tests in this PR.

`preluActivationWeights`, as defined in [fusedConv2d](https://github.com/tensorflow/tfjs/blob/fda7d7b1014ef89d61d6329e8048da0fec8ed03e/tfjs-core/src/ops/fused/conv2d.ts#L91), the shape could be similar with the input, which means it could have multiple dimensions and `dataFormat` may affect it, while the both functions do not consider `NCHW` format for it, which could be problematic:

>  @param preluActivationWeights Tensor of prelu weights to be applied as part of a `prelu` activation, typically the same shape as `x`.

In conclusion, `preluActivationWeights` deserves more research to check the correctness for `NCHW` format in WebGL, but this is unrelated to this PR, so I opened [a new ISSUE](https://github.com/tensorflow/tfjs/issues/6354) to track.
</blockquote></details>

thank you for expanding into bias and activation function, to fully support NCHW format, those are things we need to figure. SG to track them in a separate issue.

---
___
*[`tfjs-backend-webgl/src/kernels/Conv2D_impl.ts` line 261 at r1](https://reviewable.io/reviews/tensorflow/tfjs/6340#-N0WKVMr9YlliXpR2iyy:-N0mFFhw-nkPAEuy9G7x:b-y9rjob) ([raw file](https://github.com/tensorflow/tfjs/blob/bbafbf612a8705cfe5286c6bd3798ea5c992d89a/tfjs-backend-webgl/src/kernels/Conv2D_impl.ts#L261)):*
<details><summary><i>Previously, Linchenn wrote…</i></summary><blockquote>

Referring to [bias_add op in Python](https://www.tensorflow.org/api_docs/python/tf/nn/bias_add) involved in fusedConv2d, the bias is supposed to be a 1-D tensor aligned with the output channels.

I added asserts to check the shape of the bias to be a 1-D tensor or a scalar (the scalar could be broadcasted to be a 1-D tensor) and added tests to check conv2d computation with bias.
</blockquote></details>

thank you for investigating on this, this is great information.
___
*[`tfjs-backend-webgl/src/kernels/Conv2D_impl.ts` line 78 at r10](https://reviewable.io/reviews/tensorflow/tfjs/6340#-N0mCaFs5dF1l84LJWG8:-N0mCaFs5dF1l84LJWG9:b-kv18jo) ([raw file](https://github.com/tensorflow/tfjs/blob/f7de7c6a51aa8da6fb10820a8bd491fad0471e01/tfjs-backend-webgl/src/kernels/Conv2D_impl.ts#L78)):*
> ```XML
>         bias.shape.length === 0 || bias.shape[0] === convInfo.outChannels,
>         () => `WebGL conv2dByMatMul bias shape (${bias.shape}) does not ` +
>             `match the number of output channels (${convInfo.outChannels})`);
> ```

`is not compatible with the number of output channels. `
since the bias can be scalar, which does not match the channel number.

**Review by pyu10055 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6340)* status: 0 of 1 approvals obtained (waiting on @Linchenn and @qjia7)

___
*[`tfjs-backend-webgl/src/kernels/Conv2D_impl.ts` line 258 at r10](https://reviewable.io/reviews/tensorflow/tfjs/6340#-N0mG_vHCh6XlgUcuDjw:-N0mG_vHCh6XlgUcuDjx:b-b8mzzo) ([raw file](https://github.com/tensorflow/tfjs/blob/f7de7c6a51aa8da6fb10820a8bd491fad0471e01/tfjs-backend-webgl/src/kernels/Conv2D_impl.ts#L258)):*
> is not compatible with the number of output channels. 
> since the bias can be scalar, which does not match the channel number.

**Review by pyu10055 — APPROVED** [review]
> Reviewed 1 of 1 files at r3, 1 of 2 files at r4, 1 of 1 files at r6, 1 of 3 files at r8, 1 of 1 files at r9, 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6340)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @Linchenn and @qjia7)

**Review by Linchenn — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6340)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @pyu10055 and @qjia7)

*[a discussion](https://reviewable.io/reviews/tensorflow/tfjs/6340#-N0m0Gsj8c_TjvMZXbT_:-N0mHveb7G7z7k_8z93O:b-t1trzh) (no related file):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

thank you for expanding into bias and activation function, to fully support NCHW format, those are things we need to figure. SG to track them in a separate issue.
</blockquote></details>

Thank you Ping!

---
___
*[`tfjs-backend-webgl/src/kernels/Conv2D_impl.ts` line 78 at r10](https://reviewable.io/reviews/tensorflow/tfjs/6340#-N0mCaFs5dF1l84LJWG8:-N0mHtwj8RyhrOONMpj8:b-896fix) ([raw file](https://github.com/tensorflow/tfjs/blob/f7de7c6a51aa8da6fb10820a8bd491fad0471e01/tfjs-backend-webgl/src/kernels/Conv2D_impl.ts#L78)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

`is not compatible with the number of output channels. `
since the bias can be scalar, which does not match the channel number.
</blockquote></details>

Done.
___
*[`tfjs-backend-webgl/src/kernels/Conv2D_impl.ts` line 258 at r10](https://reviewable.io/reviews/tensorflow/tfjs/6340#-N0mG_vHCh6XlgUcuDjw:-N0mHtQL5XBJatax76Lo:b-896fix) ([raw file](https://github.com/tensorflow/tfjs/blob/f7de7c6a51aa8da6fb10820a8bd491fad0471e01/tfjs-backend-webgl/src/kernels/Conv2D_impl.ts#L258)):*
<details><summary><i>Previously, pyu10055 (Ping Yu) wrote…</i></summary><blockquote>

> is not compatible with the number of output channels. 
> since the bias can be scalar, which does not match the channel number.

</blockquote></details>

Done.
